system/security/securesocket/securesocket.cpp

```diff
@@ -801,132 +804,105 @@ int CSecureSocket::wait_read(unsigned timeoutms)
+            {
+                if (0 == min_size) // if min_read is 0, then whatever we have read so far is good enough (even if 0)
+                    break;
 
-void CSecureSocket::readtms(void* buf, size32_t min_size, size32_t max_size, size32_t &size_read, unsigned timeoutms)
-{
-    readTimeout(buf, min_size, max_size, size_read, timeoutms, false);
+                unsigned remainingMs = timer.remainingMs(timeoutMs);
+                rc = wait_read(remainingMs);
+                if (rc < 0)
+                    THROWJSOCKEXCEPTION_X(SOCKETERRNO(), "wait_read error");
+                if (rc == 0)
+                    THROWJSOCKEXCEPTION_X(JSOCKERR_timeout_expired, "timeout expired");
+            }
+            else
+            {
+                char errbuf[512];
+                ERR_error_string_n(ssl_err, errbuf, 512);
+                ERR_clear_error();
+                VStringBuffer errmsg("SSL_read error %d - %s", ssl_err, errbuf);
+                if (m_loglevel >= SSLogMax)
+                    DBGLOG("Warning: %s", errmsg.str());
+                THROWJSOCKEXCEPTION_X(ssl_err, errmsg);
+            }
+            // here only if nonBlocking && WANT_READ or WANT_WRITE
+            // since we do not have size_min yet, loop around and wait for more.
+        }
+    }
+    _sizeRead = sizeRead;
 }
 
 void CSecureSocket::read(void* buf, size32_t min_size, size32_t max_size, size32_t &size_read,unsigned timeoutsecs)
 {
-    readTimeout(buf, min_size, max_size, size_read, timeoutsecs, true);
+    unsigned timeoutMs = (timeoutsecs==WAIT_FOREVER) ? WAIT_FOREVER : (timeoutsecs * 1000);
+    readtms(buf, min_size, max_size, size_read, timeoutMs);
 }
 
-size32_t CSecureSocket::write(void const* buf, size32_t size)
+size32_t CSecureSocket::writetms(void const* buf, size32_t minSize, size32_t size, unsigned timeoutms)
 {
     if (size == 0)
         return 0;
-    int numwritten = SSL_write(m_ssl, buf, size);
-    // 0 is an error
-    if (numwritten <= 0)
+
+    CCycleTimer timer;
+    while (true)
     {
+        int numwritten = SSL_write(m_ssl, buf, size);
+        if (numwritten > 0)
+        {
+            // NB: minSize not used here, because not using SSL_MODE_ENABLE_PARTIAL_WRITE
+            dbgassertex(size == numwritten);
+            return numwritten;
+        }
         int err = SSL_get_error(m_ssl, numwritten);
-        // SSL_ERROR_WANT_READ/WRITE errors and retry should not be required
-        // b/c of blocking bio and SSL_MODE_ENABLE_PARTIAL_WRITE is not set
-        char errbuf[512];
-        ERR_error_string_n(err, errbuf, 512);
-        ERR_clear_error();
-        VStringBuffer errmsg("SSL_write error %d - %s", err, errbuf);
-        if (err == SSL_ERROR_ZERO_RETURN)
-            throw createJSocketException(JSOCKERR_graceful_close, errmsg);
-        else
-            throw createJSocketException(JSOCKERR_broken_pipe, errmsg);
-    }
-    return numwritten;
+        if (err != SSL_ERROR_WANT_WRITE || !nonBlocking)
```

Comment:
SSL_write() nonblocking can return SSL_WANT_READ in which case we can call wait_read() to see when to continue ...

Reply:
agreed, will change to handle either.